File: models/media_tags.go
package models

// Fork (media tags, MEDIA-TAGS-SPEC D2). The one server-side rule for a media tag: a bare
// lowercase slug. Every write path (the upload's `tags` field, PUT /api/media/:id/tags) and
// the listing's `tag` filter go through NormalizeMediaTags, strictly — an invalid tag is a
// 400, never silently dropped. (The lenient variant that drops invalid entries lives only
// where context is DERIVED rather than typed: the frontend's mediaTags.js and the one-off
// backfill script.)

import (
	"fmt"
	"regexp"
	"sort"
	"strings"
)

// MediaTagMaxLen matches the column type, VARCHAR(100)[].
const MediaTagMaxLen = 100

// ReMediaTag is a media tag after trimming and lower-folding. No `brand:` prefix: media has no
// From derivation to feed.
var ReMediaTag = regexp.MustCompile(`^[a-z0-9][a-z0-9_-]*$`)

// MediaTagInvalidError names the first tag that failed the rule. Key is the i18n key the
// handlers translate with {tag}.
type MediaTagInvalidError struct {
	Tag string
}

// MediaTagInvalidKey is the i18n key for MediaTagInvalidError.
const MediaTagInvalidKey = "media.tagInvalid"

func (e *MediaTagInvalidError) Error() string {
	return fmt.Sprintf("invalid media tag %q", e.Tag)
}

// NormalizeMediaTags trims, lower-folds, drops empties, dedupes and sorts. Any tag failing
// ReMediaTag or longer than MediaTagMaxLen is rejected with a *MediaTagInvalidError. nil and
// an empty slice both yield a non-nil empty slice, so the JSON is `[]` and the bind is '{}'.
func NormalizeMediaTags(in []string) ([]string, error) {
	seen := make(map[string]struct{}, len(in))
	out := make([]string, 0, len(in))

	for _, t := range in {
		t = strings.ToLower(strings.TrimSpace(t))
		if t == "" {
			continue
		}
		if len(t) > MediaTagMaxLen || !ReMediaTag.MatchString(t) {
			return nil, &MediaTagInvalidError{Tag: t}
		}
		if _, ok := seen[t]; ok {
			continue
		}
		seen[t] = struct{}{}
		out = append(out, t)
	}

	sort.Strings(out)
	return out, nil
}

File: frontend/src/mediaTags.js
// Fork (media tags, MEDIA-TAGS-SPEC D2/D3). Media tags are bare lowercase slugs. The server is
// always strict (models.NormalizeMediaTags rejects an invalid tag with a 400); this file holds
// the two client-side halves:
//
//   - normalizeMediaTagsLenient: for CONTEXT tags derived from the editor (a campaign's brand
//     and Tags field, a template's brand) -- invalid entries are dropped silently, never sent.
//   - isValidMediaTag: the tag inputs' before-adding check for user-typed tags, the same rule
//     the server enforces, so a bad tag is refused in place instead of failing the save.
//
// MUST MATCH models.ReMediaTag / MediaTagMaxLen in the Go tree. Disposable at v7 (D12).

export const reMediaTag = /^[a-z0-9][a-z0-9_-]*$/;
export const MEDIA_TAG_MAX_LEN = 100;

export const foldMediaTag = (t) => String(t == null ? '' : t).trim().toLowerCase();

export const isValidMediaTag = (t) => t.length <= MEDIA_TAG_MAX_LEN && reMediaTag.test(t);

// Trim, lower-fold, drop empties and invalid entries, dedupe, sort.
export const normalizeMediaTagsLenient = (tags) => {
  const out = new Set();
  (tags || []).forEach((raw) => {
    const t = foldMediaTag(raw);
    if (t && isValidMediaTag(t)) {
      out.add(t);
    }
  });
  return [...out].sort();
};
